python/lsst/ip/diffim/dipoleFitTask.py

```diff
@@ -0,0 +1,921 @@
+import numpy as np
+
+# LSST imports
+import lsst.afw.geom as afwGeom
+import lsst.afw.image as afwImage
+import lsst.meas.base as measBase
+import lsst.afw.table as afwTable
+import lsst.afw.detection as afwDet
+import lsst.pex.exceptions as pexExcept
+import lsst.pex.logging as pexLog
+import lsst.pex.config as pexConfig
+
+__all__ = ("DipoleFitPluginConfig", "DipoleFitTask", "DipoleFitPlugin",
+           "DipoleFitAlgorithm")
+
+
+# Create a new measurement task (`DipoleFitTask`) that can handle all other SFM tasks but can
+# pass a separate pos- and neg- exposure/image to the `DipoleFitPlugin`s `run()` method.
+
+
+class DipoleFitPluginConfig(measBase.SingleFramePluginConfig):
+    """
+    Class to initialize and store dipole fitting configuration parameters
+    """
+
+    maxSeparation = pexConfig.Field(
+        dtype=float, default=5.,
+        doc="Assume dipole is not separated by more than maxSeparation * psfSigma")
+
+    relWeight = pexConfig.Field(
+        dtype=float, default=0.5,
+        doc="""Relative weighting of pre-subtraction images (higher -> greater influence of pre-sub.
+        images on fit)""")
+
+    tolerance = pexConfig.Field(
+        dtype=float, default=1e-7,
+        doc="Fit tolerance")
+
+    fitBgGradient = pexConfig.Field(
+        dtype=bool, default=True,
+        doc="Include parameters to fit for linear gradient in pre-sub. images")
+
+    fitSeparateNegParams = pexConfig.Field(
+        dtype=bool, default=False,
+        doc="Include parameters to fit for negative values (flux, gradient) separately from pos.")
+
+    # Config params for classification of detected diaSources as dipole or not
+    minSn = pexConfig.Field(
+        dtype=float, default=np.sqrt(2) * 5.0,
+        doc="Minimum quadrature sum of positive+negative lobe S/N to be considered a dipole")
+
+    maxFluxRatio = pexConfig.Field(
+        dtype=float, default=0.65,
+        doc="Maximum flux ratio in either lobe to be considered a dipole")
+
+    # Choose a maxChi2DoF corresponding to a significance level of at most 0.05
+    # (note this is actually a significance not a chi2 number)
+    maxChi2DoF = pexConfig.Field(
+        dtype=float, default=0.05,
+        doc="Maximum Chi2/DoF significance of fit to be considered a dipole.")
+
+    verbose = pexConfig.Field(
+        dtype=bool, default=False,
+        doc="Be verbose; this is slow")
+
+
+class DipoleFitTask(measBase.SingleFrameMeasurementTask):
+    """This is a subclass of SingleFrameMeasurementTask which can accept
+    three input images in its run() method. Because it subclasses
+    SingleFrameMeasurementTask, and calls
+    SingleFrameMeasurementTask.run() from its run() method, it still
+    can be used identically to a standard SingleFrameMeasurementTask.
+    """
+
+    ConfigClass = DipoleFitPluginConfig
+    _DefaultName = "ip_diffim_DipoleFit"
+
+    def __init__(self, schema, algMetadata=None, dpFitPluginConfig=None, **kwds):
```

Comment:
I think you are conflating the config for the task with the config for the plugin. You use the plugin config as the config for this task AND also accept it as a separate named argument. Why? If you can use the plugin config then it is already being passed as the argument "config" and is available as "self.config" after you call SFMTask.__init__. I strongly suggest that you either use just one instance or else have separate config classes for the plugin and task.